txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
+		}
 
-	if len(t.index) == 0 {
-		return nil
-	}
-	tx := heap.Pop(t.heap).(*pricedTx)
-	delete(t.index, tx.tx.Hash)
-	return tx
-}
+		return true
+	})
 
-func (t *txPriceHeap) Contains(tx *types.Transaction) bool {
-	_, ok := t.index[tx.Hash]
-	return ok
+	return
 }
 
-// return new max-price ordered tx heap
-func newMaxTxPriceHeap() *txPriceHeap {
-	return &txPriceHeap{
-		index: make(map[types.Hash]*pricedTx),
-		heap:  newMaxTxPriceHeapImpl(),
-	}
+/* account impl */
+
+// An account is the core structure for processing
+// transactions from a specific address. The nextNonce
+// field is what separetes the enqueued from promoted:
+//
+// 	1. enqueued - transactions higher than the nextNonce
+// 	2. promoted - transactions lower than the nextNonce
+//
+// If an enqueued transaction matches the nextNonce,
+// a promoteRequest is signaled for this account
+// indicating the account's enqueued transaction(s)
+// are ready to be moved to the promoted queue.
+type account struct {
+	init               sync.Once
+	enqueued, promoted *accountQueue
+	nextNonce          uint64
 }
 
-// return new min-price ordered tx heap
-func newMinTxPriceHeap() *txPriceHeap {
-	return &txPriceHeap{
-		index: make(map[types.Hash]*pricedTx),
-		heap:  newMinTxPriceHeapImpl(),
-	}
+// Returns the next expected nonce for this account.
+func (a *account) getNonce() uint64 {
+	return atomic.LoadUint64(&a.nextNonce)
 }
 
-// Required method definitions for the standard golang heap package
-type txPriceHeapImplBase struct {
-	txs []*pricedTx
+// Sets the next expected nonce for this account.
+func (a *account) setNonce(nonce uint64) {
+	atomic.StoreUint64(&a.nextNonce, nonce)
 }
 
-func (t txPriceHeapImplBase) Len() int { return len(t.txs) }
+// Pushes the transaction onto the enqueued queue.
+func (a *account) enqueue(tx *types.Transaction, demoted bool) error {
+	a.enqueued.lock(true)
+	defer a.enqueued.unlock()
 
-func (t txPriceHeapImplBase) Swap(i, j int) {
-	t.txs[i], t.txs[j] = t.txs[j], t.txs[i]
-	t.txs[i].index = i
-	t.txs[j].index = j
-}
+	// only accept low nonce if
+	// tx was demoted
+	if tx.Nonce < a.getNonce() &&
+		!demoted {
+		return ErrNonceTooLow
+	}
 
-func (t *txPriceHeapImplBase) Push(x interface{}) {
-	n := len(t.txs)
-	job := x.(*pricedTx)
-	job.index = n
-	t.txs = append(t.txs, job)
-}
+	// enqueue tx
+	a.enqueued.push(tx)
 
-func (t *txPriceHeapImplBase) Pop() interface{} {
-	old := *t
-	n := len(old.txs)
-	job := old.txs[n-1]
-	job.index = -1
-	t.txs = old.txs[0 : n-1]
-	return job
+	return nil
 }
 
-func (t txPriceHeapImplBase) Less(i, j int) bool {
-	return i < j
-}
+// Promote moves eligible transactions from enqueued to promoted.
+//
+// Eligible transactions are all sequential in order of nonce
+// and the first one has to have nonce less (or equal) to the account's
+// nextNonce.
+func (a *account) promote() uint64 {
+	a.promoted.lock(true)
+	a.enqueued.lock(true)
+	defer func() {
+		a.promoted.unlock()
+		a.enqueued.unlock()
```

Comment:
Please reverse the order of unlocking, as this can lead to a deadlock (free the enqueued lock first, then the promoted one)

Reply:
[Reversed](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/65dd59da6bd4b4b884127e3a5d08d5e18ae0375a)